File: SimboDev24/sample/javascript/sample02/Tennis.js
const hitsToScore = new Map();

hitsToScore.set(0, "Love");
hitsToScore.set(1, "Fifteen");
hitsToScore.set(2, "Thirty");
hitsToScore.set(3, "Forty");

function Tennis(aName, bName) {

    this.aName = aName;
    this.bName = bName;

    this.aHits = 0;
    this.bHits = 0;

    this.a = function () {
        this.aHits++;
    };

    this.b = function () {
        this.bHits++;
    };

    this.diffMoreThanOne = function () {
        return Math.abs(this.aHits - this.bHits) > 1;
    };

    this.sameScore = function () {
        return this.aHits === this.bHits;
    };

    this.afterDeuce = function () {

        return Math.max(this.aHits, this.bHits) > 3;
    };

    this.score = function () {


        if (this.sameScore()) {
            return this.aHits >= 3
                ? 'Deuce'
                : hitsToScore.get(this.aHits) + ' All';
        } else {
            if (this.afterDeuce()) {

                return this.leadingPlayer()
                    +
                    (this.diffMoreThanOne()
                        ? ' Wins'
                        : ' Adv.');

            } else {
                return hitsToScore.get(this.aHits) + ' ' + hitsToScore.get(this.bHits);
            }
        }


    };


    this.leadingPlayer = function () {
        return this.aHits > this.bHits
            ? this.aName
            : this.bName;
    }


}

module.exports = Tennis;

File: SimboDev24/sample/java/01/simbodev/src/main/java/kuma/Tennis.java
package kuma;

import java.util.HashMap;
import java.util.Map;

public class Tennis {

    private static Map<Integer, String> scoreToShow = new HashMap<>();

    static {
        scoreToShow.put(0, "Love");
        scoreToShow.put(1, "Fifteen");
        scoreToShow.put(2, "Thirty");
        scoreToShow.put(3, "Forty");
    }

    private int aScore;
    private int bScore;
    private String aName;
    private String bName;

    public Tennis(String aName, String bName) {
        this.aName = aName;
        this.bName = bName;
    }

    public String score() {

        if (sameScore()) {
            return aScore >= 3
                    ? "Deuce"
                    : scoreToShow.get(aScore) + " All";
        }

        if (Math.max(aScore, bScore) > 3) {
            return isWin()
                    ? leadingPlayer() + " Wins"
                    : leadingPlayer() + " Adv";
        }
        return scoreToShow.get(aScore) + " " + scoreToShow.get(bScore);


    }

    private boolean sameScore() {
        return aScore == bScore;
    }

    private boolean isWin() {
        return Math.abs(bScore - aScore) > 1;
    }

    private String leadingPlayer() {
        return aScore > bScore
                ? aName
                : bName;
    }

    public void a() {
        aScore++;
    }

    public void b() {
        bScore++;
    }
}